### PR TITLE
`no_std` support for the `url` crate with no breaking changes for feature `std`

### PR DESCRIPTION
**Pull request by mspiegel**
Continuing the efforts from #831, this povides no_std support for the url crate. The default features of the url crate are changed from [ ] to [ 'std' ]. Changing the default features is a breaking change. The recommendation is to increase the major version of the url crate. The breaking change affects users who have enabled `default-features = false`. A section to the crate docs has been added with an explanation for affected users. Affected users will need to enable either the 'std' or 'alloc' feature. The MSRV is unchanged for the 'std' feature.

### CODE REVIEW

**Comment by domenukk** [comment]
I just tried your pr. I like the idea to have an extra `std_core_compat`.
However, note your PR currently isn't really no_std, as the dependencies are not no_std.

Take a look at the Cargo.toml changes / dependency changes in PR #831 and/or look at the CI workflow for how to test this.

TL;DR:

```rust
cargo +nightly check -Zbuild-std=core,alloc --target aarch64-unknown-none -v --release --no-default-features --features=alloc
```

**Comment by Manishearth** [comment]
> Changing the default features is a breaking change

I don't think this is always considered the case, at least not in practice in the Rust community. Default features are regularly used to introduce  changes like this in a non-breaking way.

(In the long run there are some potential features that could be added to rust that would make things like this no longer break builds)

**Comment by mspiegel** [comment]
> I just tried your pr. I like the idea to have an extra `std_core_compat`. However, note your PR currently isn't really no_std, as the dependencies are not no_std.
> 
> Take a look at the Cargo.toml changes / dependency changes in PR #831 and/or look at the CI workflow for how to test this.
> 
> TL;DR:
> 
> ```rust
> cargo +nightly check -Zbuild-std=core,alloc --target aarch64-unknown-none -v --release --no-default-features --features=alloc
> ```

Thanks! I pulled in the changes to .github/workflows/main.yml from #831. The workflow is awaiting approval before it will run. In the meantime, you can see the results of the workflow here: https://github.com/mspiegel/foobar/pull/1

**Comment by mspiegel** [comment]
> > Changing the default features is a breaking change
> 
> I don't think this is always considered the case, at least not in practice in the Rust community. Default features are regularly used to introduce changes like this in a non-breaking way.
> 
> (In the long run there are some potential features that could be added to rust that would make things like this no longer break builds)

The new section to the crate docs that explains the default features has been updated to specify 2.5.2 as the last version where the default features of the url crate are empty. If the version bump is major or minor, the docs text will be correct either way.

**Comment by mspiegel** [comment]
Closing in favor of #831
